threedi_models_and_simulations/widgets/simulation_wizard.py

```diff
@@ -681,6 +696,66 @@ def connect_signals(self):
         self.pb_upload_2d_laterals.clicked.connect(partial(self.load_csv, self.TYPE_2D))
         self.cb_2d_interpolate.stateChanged.connect(partial(self.interpolate_changed, self.TYPE_2D))
 
+    def setup_substance_concentrations(self):
+        if hasattr(self, "groupbox"):
+            self.groupbox.setParent(None)
+        if not self.substances:
+            return
+        substance_concentration_widget = SubstanceConcentrationsWidget(self.substances, self.handle_substance_timesteps)
+        self.groupbox = substance_concentration_widget.groupbox
+        self.substance_concentrations = substance_concentration_widget.substance_concentrations
+        parent_layout = self.layout()
+        parent_layout.addWidget(self.groupbox, 3, 2)
+
+    def handle_substance_header(self, substance_list):
+        """
+        Fetch first csv row and handle potential header.
+        Return None if fetch successful or error message if file is empty or have invalid structure.
+        """
+        error_message = None
+        if not substance_list:
+            error_message = "Substance concentration list is empty!"
+            return error_message
+        header = substance_list[0]
+        if len(header) != 2:
+            error_message = "Wrong csv format for substance concentrations!"
+        if error_message is None:
+            try:
+                timeseries_candidate = header[-1]
+                [[float(f) for f in line.split(",")] for line in timeseries_candidate.split("\n")]
```

Comment:
Same logic repeats in a line 750. Please move the logic to separate static method or utils function.